README.md
# Prime-Sieve-Dart

A prime number generator written in Dart, using Euler's sieve and a wheel size
of 210.

The code is based as closely as possible on
[hacatu](https://github.com/hacatu)'s
[Prime-Sieve](https://github.com/hacatu/Prime-Sieve).

## Performance

The performance of the prime determination is about half as fast as the C
implementation. On my machine, it requires about 7.5&#8239;s (Dart) instead of 3.5&#8239;s
(C) to find all primes up to 10⁹.

(Printing, however, is way slower, instead of 3.5&#8239;s for the C version, the
output takes about 80&#8239;s.)

### More performance or flexibility?

If you need more performance (about 20x) or flexibility (e.g., iterators),
[Kim Walisch's `primesieve` C/C++ library](https://github.com/kimwalisch/primesieve)
is the way to go. It also may come
[pre-packaged with/for your OS](https://github.com/kimwalisch/primesieve#installation).

## Usage
`dart example/prime_sieve.dart [options] <max>`

Determine primality for all numbers up to `max`. `max` should not exceed
about 2⁶⁰ (or 2⁵⁰, when translating to JavaScript) or the size of your
memory, whatever is smaller.

## Options
* `--help`: Print help
* `--[no-]count`: Whether to output the count of primes found.
  (Default: true.)
* `--[no-]print`: Whether to output the primes after calculating them.
  (Default: true.)
* `--query`: Whether to enter a query input loop for primality tests after
  possibly printing the values.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

# 0.9.2 - 2020-02-12
## Added

## Fixed

## Changed
* Executable now called `primzahlsieb` to make names more unique
* Example renamed to `example/main.dart` (for uniqueness as well)

# 0.9.1 - 2020-02-12
## Added
* Added `prime_sieve_dart` as an executable (instead of an example)
* Added minimalistic documentation example

## Fixed

## Changed
* Changed package name to avoid near-conflict


# 0.9.0 - 2020-01-20
## Added
* API documentation
* Changelog

## Fixed

## Changed
* Executable is now in `example/main.dart`.
* Ran `dartfmt` over all the files (with the exception of `_offsets` in
  `lib/prime_sieve.dart`, which would really look ugly).


# 0.0.1 - 2020-01-19
## Added

## Fixed

## Changed
* [Ported from hakatu's C code](https://github.com/hacatu/Prime-Sieve)


ptest.c
/* ptest.c
 * test the Prime Sieve
*/
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include "sieve210.h"


int main(int argc, char** argv){
	uint64_t max = 1000000;
	if(argc > 1){
		max = atoll(argv[1]);
	}else{
		puts("No maximum specified.  Assuming \"ptest 1000000\".");
	}
	uint64_t* P = sievePrimes(max);
	if(argc > 2){
		char print = atoi(argv[2]);
		if(print){
		    uint64_t max = P[0], i;
			for(i = 1; i <= max; ++i){
				printf("%lu\n",P[i]);
			}
		}
	}else{
		printf("Printing unspecified.  Use \"1\" to print and \"0\" to not print.\n\
Running in primality testing mode.  Enter a number up to %lu inclusive:\n", max);
		char* line = 0;
		size_t len = 0;
		while(1 < getline(&line, &len, stdin)){
			if(isPrime(atoll(line),P)){
				puts("prime");
			}else{
				puts("not prime");
			}
		}
		free(line);
	}
	free(P);
}


lib/sieve210.dart
import 'dart:typed_data';
import 'dart:math';

//the spaces between the wheel spokes.  The sieve skips all multiples of 2, 3, 5, and 7
//and that can be done by starting at 1 and adding these numbers in sequence.
class Sieve210 {
	static final _offsets = [
		10, 2, 4, 2, 4, 6,  2, 6,
		4, 2, 4, 6, 6, 2,  6, 4,
		2, 6, 4, 6, 8, 4,  2, 4,
		2, 4, 8, 6, 4, 6,  2, 4,
		6, 2, 6, 6, 4, 2,  4, 6,
		2, 6, 4, 2, 4, 2, 10, 2
	];

	//a 210 entry array that will be initialized with the index in the wheel
  // (offsets) for each modulus 210 not a multiple of 2, 3, 5, or 7
	static List<int> _lut_offsets;

	//a function that calculates the index of a number in the is_composite array.
	//It will be the xth number relatively prime to 210
  static int wheelSpokes(int n) {
    return n ~/ 210 * 48 + _lut_offsets[n % 210];
  }

  Uint64List primes;
  int length;

  _init_lut_offsets() {
    // The last two elements are not used; makes the code simpler, however.
    var l = List<int>(212);
    var key = 1;
    var value = 0;
    l[key] = value;
    for (var o in _offsets) {
      key += o;
      value++;
      l[key] = value;
    }
    // Allow some parallelism
    _lut_offsets = l;
  }

	//sievePrimes returns a pointer to an array of all the primes up to and including max.
	//the 0 element of this array is the length of the array.
  Sieve210(int max) {
    if (_lut_offsets == null) _init_lut_offsets();
    //create a list of chars representing whether or not every number to be sieved is composite.  Initialized to 0.
    var is_composite = Uint8List(max ~/ 210 * 48 + 48);

    //primes_s is the estimated number of primes up to max, plus 1 (for the length of the primes array stored at 0)
    //TODO: calculate this better by using logarithmic integrals or by only pushing and sieving up to sqrt(max), then pushing.
    var primes_s = (max / log(max+1) * 1.2 + 10).ceil();
    primes = Uint64List(primes_s);

    //initializes the primes which are part of the wheel.
    primes[0] = 2;
    primes[1] = 3;
    primes[2] = 5;
    primes[3] = 7;

    //declare variables.
    int i = 4, p1, p2, i1 = 1, i2, step1 = 0, step2, multiple;
    int sqrtmax = sqrt(max).ceil();

    //for every number still marked prime, cross off its multiples with every other number still marked prime and add it to the primes list.
    for (p1 = 11; p1 <= sqrtmax; ++i1, p1 += _offsets[step1]) {
      //switch to the next step in offsets, wrapping around if it overflows.
      ++step1;
      if (step1 == 48) {
        step1 = 0;
      }

      //skip composite numbers.  Note that is_composite[i1] is just looking up if i1 has been marked composite, not doing a mathematical test.
      if (is_composite[i1] != 0) {
        continue;
      }

      //add p1 to the list of primes
      primes[i++] = p1;
      if (i == primes_s) {
        //re allocate the primes array.  TODO: do this at most once, either using li instead of log or by stopping at the sqrt and allocating what's left minus the number of numbers crossed off.
        primes_s = (primes_s * 1.1).ceil();
        var t = Uint64List(primes_s);
        for (var c = 0; c < i; c++) {
          t[c] = primes[c];
        }
        primes = t;
      }

      //mark all multiples of p1 and some number still marked as prime as composite.
      //note: if a number is being marked composite, it times p1 still needs to be crossed of, e.g. p1 = 11, p2 = 11, cross off 121 and 1331 and so on.
      i2 = i1;
      step2 = step1 - 1;
      for (p2 = p1; p2 <= max ~/ p1; ++i2, p2 += _offsets[step2]) {
        ++step2;
        if (step2 == 48) {
          step2 = 0;
        }

        if (is_composite[i2] != 0) {
          continue;
        }

        multiple = p1 * p2;
        do {
          //cross off all multiple = p1*p2^n < max
          is_composite[wheelSpokes(multiple)] = 1;
        } while ((multiple *= p1) <= max);
      }
    }

    while (p1 <= max) {
      ++step1;
      if (step1 == 48) {
        step1 = 0;
      }
      if (is_composite[i1] == 0) {
        //add p1 to the list of primes
        primes[i++] = p1;
        if (i == primes_s) {
          //re allocate the primes array.  TODO: do this at most once, either using li instead of log or by stopping at the sqrt and allocating what's left minus the number of numbers crossed off.
          primes_s = (primes_s * 1.1).ceil();
          var t = Uint64List(primes_s);
          for (var c = 0; c < i; c++) {
            t[c] = primes[c];
          }
          primes = t;
        }

      }
      ++i1;
      p1 += _offsets[step1];
    }

    length = i;
  }

	//test if a number is prime by checking if it is in a list of primes.
	//binary search primes for n
	//TODO: use a jumping search
  bool isPrime(int n) {
    //start with max as the number of primes in primes.
    var max = primes.length;
    //start with min as 0 and mid as the average of min and max.
    var min = 0, mid = max ~/ 2;
    //binary search
    while (n != primes[mid]) {
      //eliminate the left side.
      if (n > primes[mid]) {
        min = mid + 1;
      } else {
        //eliminate the right side
        max = mid;
      }
      mid = max ~/ 2 + min ~/ 2;
      if (max == min) {
        //everything has been searched
        return false;
      }
    }
    return true;
  }
}


lib/prime_sieve.dart
import 'dart:typed_data';
import 'dart:math';

/// Calculate prime numbers up to [max].
///
/// Uses a sieve with wheel of size 210.
//the spaces between the wheel spokes.  The sieve skips all multiples of 2, 3, 5, and 7
//and that can be done by starting at 1 and adding these numbers in sequence.
class Sieve210 {
  /// The distances of the spokes in the wheel.
  static final _offsets = [
    10, 2, 4, 2, 4, 6, 2, 6,
    4, 2, 4, 6, 6, 2, 6, 4,
    2, 6, 4, 6, 8, 4, 2, 4,
    2, 4, 8, 6, 4, 6, 2, 4,
    6, 2, 6, 6, 4, 2, 4, 6,
    2, 6, 4, 2, 4, 2, 10, 2
  ];

  /// A 210 entry array that will be initialized with the index in the wheel
  /// (offsets) for each modulus 210 not a multiple of 2, 3, 5, or 7.
  static var _lut_offsets = _init_lut_offsets();

  /// Calculate the index of a number in the is_composite array.
  ///
  /// It will be the xth number relatively prime to 210.
  static int _wheelSpokes(int n) {
    return n ~/ 210 * 48 + _lut_offsets[n % 210];
  }

  /// Calculated primes are stored here.
  Uint64List primes;
  /// Number of primes in [primes].
  int length;

  /// Create the [_lut_offsets] table based on [_offsets].
  static _init_lut_offsets() {
    // The last two elements are not used; makes the code simpler, however.
    var l = List<int>(212);
    var key = 1;
    var value = 0;
    l[key] = value;
    for (var o in _offsets) {
      key += o;
      value++;
      l[key] = value;
    }
    return l;
  }

  /// Create the list of primes up to [max].
  ///
  /// Results are stored in [primes] (the primes themselves) and [length]
  /// (the number of items found).
  //sievePrimes returns a pointer to an array of all the primes up to and including max.
  //the 0 element of this array is the length of the array.
  Sieve210(int max) {
    //create a list of chars representing whether or not every number to be sieved is composite.  Initialized to 0.
    var is_composite = Uint8List(max ~/ 210 * 48 + 48);

    //primes_s is the estimated number of primes up to max, plus 1 (for the length of the primes array stored at 0)
    //TODO: calculate this better by using logarithmic integrals or by only pushing and sieving up to sqrt(max), then pushing.
    var primes_s = (max / log(max + 1) * 1.2 + 10).ceil();
    primes = Uint64List(primes_s);

    //initializes the primes which are part of the wheel.
    primes[0] = 2;
    primes[1] = 3;
    primes[2] = 5;
    primes[3] = 7;

    //declare variables.
    int i = 4, p1, p2, i1 = 1, i2, step1 = 0, step2, multiple;

    //for every number still marked prime, cross off its multiples with every other number still marked prime and add it to the primes list.
    for (p1 = 11; p1 <= max; ++i1, p1 += _offsets[step1]) {
      //switch to the next step in offsets, wrapping around if it overflows.
      ++step1;
      if (step1 == 48) {
        step1 = 0;
      }

      //skip composite numbers.  Note that is_composite[i1] is just looking up if i1 has been marked composite, not doing a mathematical test.
      if (is_composite[i1] != 0) {
        continue;
      }

      //add p1 to the list of primes
      primes[i++] = p1;
      if (i == primes_s) {
        //re allocate the primes array.  TODO: do this at most once, either using li instead of log or by stopping at the sqrt and allocating what's left minus the number of numbers crossed off.
        primes_s = (primes_s * 1.1).ceil();
        var t = Uint64List(primes_s);
        for (var c = 0; c < i; c++) {
          t[c] = primes[c];
        }
        primes = t;
      }

      //mark all multiples of p1 and some number still marked as prime as composite.
      //note: if a number is being marked composite, it times p1 still needs to be crossed of, e.g. p1 = 11, p2 = 11, cross off 121 and 1331 and so on.
      i2 = i1;
      step2 = step1 - 1;
      for (p2 = p1; p2 <= max ~/ p1; ++i2, p2 += _offsets[step2]) {
        ++step2;
        if (step2 == 48) {
          step2 = 0;
        }

        if (is_composite[i2] != 0) {
          continue;
        }

        multiple = p1 * p2;
        do {
          //cross off all multiple = p1*p2^n < max
          is_composite[_wheelSpokes(multiple)] = 1;
        } while ((multiple *= p1) <= max);
      }
    }

    length = i;
  }

  /// Check whether the parameter is present in the list of [primes].
  //test if a number is prime by checking if it is in a list of primes.
  //binary search primes for n
  //TODO: use a jumping search
  bool isPrime(int n) {
    //start with max as the number of primes in primes.
    var max = length;
    //start with min as 0 and mid as the average of min and max.
    var min = 0, mid = max ~/ 2;
    //binary search
    while (n != primes[mid]) {
      //eliminate the left side.
      if (n > primes[mid]) {
        min = mid + 1;
      } else {
        //eliminate the right side
        max = mid;
      }
      mid = max ~/ 2 + min ~/ 2;
      if (max == min) {
        //everything has been searched
        return false;
      }
    }
    return true;
  }
}


lib/prime_sieve_dart.dart
import 'dart:typed_data';
import 'dart:math';

/// Calculate prime numbers up to [max].
///
/// Uses a sieve with wheel of size 210.
//the spaces between the wheel spokes.  The sieve skips all multiples of 2, 3, 5, and 7
//and that can be done by starting at 1 and adding these numbers in sequence.
class Sieve210 {
  /// The distances of the spokes in the wheel.
  static final _offsets = [
    10, 2, 4, 2, 4, 6, 2, 6,
    4, 2, 4, 6, 6, 2, 6, 4,
    2, 6, 4, 6, 8, 4, 2, 4,
    2, 4, 8, 6, 4, 6, 2, 4,
    6, 2, 6, 6, 4, 2, 4, 6,
    2, 6, 4, 2, 4, 2, 10, 2
  ];

  /// A 210 entry array that will be initialized with the index in the wheel
  /// (offsets) for each modulus 210 not a multiple of 2, 3, 5, or 7.
  static final _lut_offsets = _init_lut_offsets();

  /// Calculate the index of a number in the is_composite array.
  ///
  /// It will be the xth number relatively prime to 210.
  static int _wheelSpokes(int n) {
    return n ~/ 210 * 48 + _lut_offsets[n % 210];
  }

  /// Calculated primes are stored here.
  Uint64List primes;
  /// Number of primes in [primes].
  int length;

  /// Create the [_lut_offsets] table based on [_offsets].
  static _init_lut_offsets() {
    // The last two elements are not used; makes the code simpler, however.
    var l = List<int>(212);
    var key = 1;
    var value = 0;
    l[key] = value;
    for (var o in _offsets) {
      key += o;
      value++;
      l[key] = value;
    }
    return l;
  }

  /// Create the list of primes up to [max].
  ///
  /// Results are stored in [primes] (the primes themselves) and [length]
  /// (the number of items found).
  //sievePrimes returns a pointer to an array of all the primes up to and including max.
  //the 0 element of this array is the length of the array.
  Sieve210(int max) {
    //create a list of chars representing whether or not every number to be sieved is composite.  Initialized to 0.
    var is_composite = Uint8List(max ~/ 210 * 48 + 48);

    //primes_s is the estimated number of primes up to max, plus 1 (for the length of the primes array stored at 0)
    //TODO: calculate this better by using logarithmic integrals or by only pushing and sieving up to sqrt(max), then pushing.
    var primes_s = (max / log(max + 1) * 1.2 + 10).ceil();
    primes = Uint64List(primes_s);

    //initializes the primes which are part of the wheel.
    primes[0] = 2;
    primes[1] = 3;
    primes[2] = 5;
    primes[3] = 7;

    //declare variables.
    int i = 4, p1, p2, i1 = 1, i2, step1 = 0, step2, multiple;

    //for every number still marked prime, cross off its multiples with every other number still marked prime and add it to the primes list.
    for (p1 = 11; p1 <= max; ++i1, p1 += _offsets[step1]) {
      //switch to the next step in offsets, wrapping around if it overflows.
      ++step1;
      if (step1 == 48) {
        step1 = 0;
      }

      //skip composite numbers.  Note that is_composite[i1] is just looking up if i1 has been marked composite, not doing a mathematical test.
      if (is_composite[i1] != 0) {
        continue;
      }

      //add p1 to the list of primes
      primes[i++] = p1;
      if (i == primes_s) {
        //re allocate the primes array.  TODO: do this at most once, either using li instead of log or by stopping at the sqrt and allocating what's left minus the number of numbers crossed off.
        primes_s = (primes_s * 1.1).ceil();
        var t = Uint64List(primes_s);
        for (var c = 0; c < i; c++) {
          t[c] = primes[c];
        }
        primes = t;
      }

      //mark all multiples of p1 and some number still marked as prime as composite.
      //note: if a number is being marked composite, it times p1 still needs to be crossed of, e.g. p1 = 11, p2 = 11, cross off 121 and 1331 and so on.
      i2 = i1;
      step2 = step1 - 1;
      for (p2 = p1; p2 <= max ~/ p1; ++i2, p2 += _offsets[step2]) {
        ++step2;
        if (step2 == 48) {
          step2 = 0;
        }

        if (is_composite[i2] != 0) {
          continue;
        }

        multiple = p1 * p2;
        do {
          //cross off all multiple = p1*p2^n < max
          is_composite[_wheelSpokes(multiple)] = 1;
        } while ((multiple *= p1) <= max);
      }
    }

    length = i;
  }

  /// Check whether the parameter is present in the list of [primes].
  //test if a number is prime by checking if it is in a list of primes.
  //binary search primes for n
  //TODO: use a jumping search
  bool isPrime(int n) {
    //start with max as the number of primes in primes.
    var max = length;
    //start with min as 0 and mid as the average of min and max.
    var min = 0, mid = max ~/ 2;
    //binary search
    while (n != primes[mid]) {
      //eliminate the left side.
      if (n > primes[mid]) {
        min = mid + 1;
      } else {
        //eliminate the right side
        max = mid;
      }
      mid = max ~/ 2 + min ~/ 2;
      if (max == min) {
        //everything has been searched
        return false;
      }
    }
    return true;
  }
}


sieve210.c
//sieve210.c
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include "sieve210.h"

//the spaces between the wheel spokes.  The sieve skips all multiples of 2, 3, 5, and 7
//and that can be done by starting at 1 and adding these numbers in sequence.
const char offsets[] = {
	10, 2, 4, 2, 4, 6,  2, 6,
	 4, 2, 4, 6, 6, 2,  6, 4,
	 2, 6, 4, 6, 8, 4,  2, 4,
	 2, 4, 8, 6, 4, 6,  2, 4,
	 6, 2, 6, 6, 4, 2,  4, 6,
	 2, 6, 4, 2, 4, 2, 10, 2
};

//a 210 char array that has the index in the wheel (offsets) for each modulus 210 not a multiple of 2, 3, 5, or 7
const char lut_offsets[] = {
	[  1] =  0, [ 11] =  1, [ 13] =  2, [ 17] =  3, [ 19] =  4, [ 23] =  5, [ 29] =  6, [ 31] =  7,
	[ 37] =  8, [ 41] =  9, [ 43] = 10, [ 47] = 11, [ 53] = 12, [ 59] = 13, [ 61] = 14, [ 67] = 15,
	[ 71] = 16, [ 73] = 17, [ 79] = 18, [ 83] = 19, [ 89] = 20, [ 97] = 21, [101] = 22, [103] = 23,
	[107] = 24, [109] = 25, [113] = 26, [121] = 27, [127] = 28, [131] = 29, [137] = 30, [139] = 31,
	[143] = 32, [149] = 33, [151] = 34, [157] = 35, [163] = 36, [167] = 37, [169] = 38, [173] = 39,
	[179] = 40, [181] = 41, [187] = 42, [191] = 43, [193] = 44, [197] = 45, [199] = 46, [209] = 47
};

//a function that calculates the index of a number in the is_composite array.
//It will be the xth number relatively prime to 210
uint64_t wheelSpokes(uint64_t n){
	return n/210*48 + lut_offsets[n%210];
}

//sievePrimes returns a pointer to an array of all the primes up to and including max.
//the 0 element of this array is the length of the array.
uint64_t* sievePrimes(uint64_t max){
	//create a list of chars representing whether or not every number to be sieved is composite.  Initialized to 0.
    char* is_composite = calloc(max/210*48 + 48, sizeof(char));
    if(!is_composite){//return a null pointer if allocation failed.
		return 0;
	}
	
	//primes_s is the estimated number of primes up to max, plus 1 (for the length of the primes array stored at 0)
	//TODO: calculate this better by using logarithmic integrals or by only pushing and sieving up to sqrt(max), then pushing.
    uint64_t primes_s = max/log(max) + 1;
    uint64_t* primes = malloc(primes_s*sizeof(uint64_t));
    if(!primes){
    	free(is_composite);
		return 0;
	}
	
	//initializes the primes which are part of the wheel.
	primes[1] = 2;
	primes[2] = 3;
	primes[3] = 5;
	primes[4] = 7;
	
	//declare variables.
    uint64_t i = 5, p1, p2, i1, i2, step1, step2, multiple;
    uint64_t* t;
    
    //for every number still marked prime, cross off its multiples with every other number still marked prime and add it to the primes list.
    for(i1 = 1, p1 = 11, step1 = 0; p1 <= max; ++i1, p1 += offsets[step1]){
    	//switch to the next step in offsets, wrapping around if it overflows.
		++step1;
		if(step1 == 48){
			step1 = 0;
		}
		
		//skip composite numbers.  Note that is_composite[i1] is just looking up if i1 has been marked composite, not doing a mathematical test.
        if(is_composite[i1]){
			continue;
		}
		
		//add p1 to the list of primes
        primes[i++] = p1;
        if(i == primes_s){//re allocate the primes array.  TODO: do this at most once, either using li instead of log or by stopping at the sqrt and allocating what's left minus the number of numbers crossed off.
			primes_s *= 1.1;
			t = realloc(primes, primes_s*sizeof(uint64_t));
			if(!t){
				free(is_composite);
				return primes;
			}
			primes = t;
		}
		
		//mark all multiples of p1 and some number still marked as prime as composite.
		//note: if a number is being marked composite, it times p1 still needs to be crossed of, e.g. p1 = 11, p2 = 11, cross off 121 and 1331 and so on.
        for(i2 = i1, p2 = p1, step2 = step1; p2 <= max/p1; ++i2, p2 += offsets[step2]){
			++step2;
			if(step2 == 48){
				step2 = 0;
			}
			
            if(is_composite[i2]){
				continue;
			}
			
            multiple = p1*p2;
            do{//cross off all multiple = p1*p2^n < max
                is_composite[wheelSpokes(multiple)] = 1;
            }while((multiple *= p1) <= max);
        }
    }
    
    //is_composite is no longer needed
    free(is_composite);
    primes[0] = i - 1;//set primes[0] to the number of primes in primes.
    t = realloc(primes, i*sizeof(uint64_t));//shrink primes to the proper length.  TODO: make primes only ever need one resize.
    if(t){
		primes = t;
	}
    return primes;
}


//test if a number is prime by checking if it is in a list of primes.
//binary search primes for n
//TODO: use a jumping search
char isPrime(uint64_t n, const uint64_t* primes){
	//start with max as the number of primes in primes.
	uint64_t max = primes[0];
	//shift primes to eliminate the length from the list.
	primes += 1;
	//start with min as 0 and mid as the average of min and max.
	uint64_t min = 0, mid = max/2;
	//binary search
	while(n != primes[mid]){
		//eliminate the left side.
		if(n > primes[mid]){
			min = mid + 1;
		}else{//eliminate the right side
			max = mid;
		}
		mid = max/2 + min/2;
		if(max == min){//everything has been searched
			return 0;
		}
	}
	return 1;
}

